main/vfs_http.c

```diff
@@ -0,0 +1,791 @@
+    params->type = VFS_HTTP_TYPE_MOUNT;
+  } else if (path[strlen(path) - 1] == '/') {
+    params->path[strlen(params->path) - 1] = '\0';
+    params->type = VFS_HTTP_TYPE_DIRECTORY;
+  } else {
+    params->type = VFS_HTTP_TYPE_FILE;
+  }
+
+  return 0;
+}
+
+static int vfs_http_error(const char *op, const char *path)
+{
+  switch(errno) {
+    case ENOENT:
+      LOG_WARN("%s %s: %s", op, path, strerror(errno));
+      return HTTP_NOT_FOUND;
+
+    case ENOTSUP:
+      LOG_WARN("%s %s: %s", op, path, strerror(errno));
+      return HTTP_METHOD_NOT_ALLOWED;
+
+    case EACCES:
+      LOG_WARN("%s %s: %s", op, path, strerror(errno));
+      return HTTP_FORBIDDEN;
+
+    default:
+      LOG_ERROR("%s %s: %s", op, path, strerror(errno));
+      return HTTP_INTERNAL_SERVER_ERROR;
+  }
+}
+
+struct vfs_http_stat {
+  size_t size;
+  time_t mtime;
+};
+
+static int vfs_http_stat(const struct vfs_http_params *params, struct vfs_http_stat *s)
+{
+  struct stat st = {};
+  int err = 0;
+
+  if (stat(params->path, &st)) {
+    return vfs_http_error("stat", params->path);
+  }
+
+  s->size = st.st_size;
+  s->mtime = st.st_mtime;
+
+  return err;
+}
+
+static int vfs_http_open(FILE **filep, const struct vfs_http_params *params, const char *mode)
+{
+  if (params->type != VFS_HTTP_TYPE_FILE) {
+    return HTTP_UNPROCESSABLE_ENTITY;
+  }
+
+  if ((*filep = fopen(params->path, mode))) {
+    return 0;
+  } else {
+    return vfs_http_error("fopen", params->path);
+  }
+}
+
+static int vfs_http_opendir(struct vfs_http_params *params)
+{
+  if (params->type != VFS_HTTP_TYPE_DIRECTORY && params->type != VFS_HTTP_TYPE_MOUNT) {
+    return HTTP_UNPROCESSABLE_ENTITY;
+  }
+
+  if ((params->dir = opendir(params->path))) {
+    return 0;
+  } else {
+    return vfs_http_error("opendir", params->path);
+  }
+}
+
+static int vfs_http_mkdir(const struct vfs_http_params *params)
+{
+  if (params->type != VFS_HTTP_TYPE_DIRECTORY) {
+    return HTTP_UNPROCESSABLE_ENTITY;
+  }
+
+  LOG_INFO("%s", params->path);
+
+  if (mkdir(params->path, VFS_HTTP_MKDIR_MODE)) {
+    return vfs_http_error("mkdir", params->path);
+  }
+
+  return 0;
+}
+
+static int vfs_http_rmdir(const struct vfs_http_params *params)
+{
+  if (params->type != VFS_HTTP_TYPE_DIRECTORY) {
+    return HTTP_UNPROCESSABLE_ENTITY;
+  }
+
+  LOG_INFO("%s", params->path);
+
+  if (rmdir(params->path)) {
+    return vfs_http_error("rmdir", params->path);
+  }
+
+  return 0;
+}
+
+static int vfs_http_unlink(const struct vfs_http_params *params)
+{
+  if (params->type != VFS_HTTP_TYPE_FILE) {
+    return HTTP_UNPROCESSABLE_ENTITY;
+  }
+
+  LOG_INFO("%s", params->path);
+
+  if (unlink(params->path)) {
+    return vfs_http_error("unlink", params->path);
+  }
+
+  return 0;
+}
+
+static int vfs_copy(FILE *read, FILE *write)
+{
+  int c;
+  size_t size = 0;
+
+  LOG_DEBUG("read=%p, write=%p", read, write);
+
+  while ((c = fgetc(read)) != EOF) {
+    if (fputc(c, write) == EOF) {
```

Comment:
Needs some benchmarking to see if this can be optimized to improve file upload/download performance.

Reply:
https://github.com/qmsk/esp/issues/51